src/lib/blkalloc/varsize_blk_allocator.cpp

```diff
@@ -614,10 +616,14 @@ blk_count_t VarsizeBlkAllocator::alloc_blks_direct(blk_count_t nblks, blk_alloc_
                 cur_blk_id = b.start_bit + b.nbits;
             }
         }
-        if (++portion_num == get_num_portions()) { portion_num = 0; }
-        BLKALLOC_LOG(TRACE, "alloc direct unable to find in prev portion, searching in portion={}, start_portion={}",
-                     portion_num, m_start_portion_num);
-    } while (nblks_remain && (portion_num != m_start_portion_num) && !hints.is_contiguous && out_blkid.has_room());
+        if (nblks_remain) {
+            auto curr_portion = portion_num;
+            if (++portion_num == get_num_portions()) { portion_num = 0; }
+            BLKALLOC_LOG(
+                TRACE, "alloc direct unable to find in curr portion {}, will searching in portion={}, start_portion={},continue={}, out_blkid num_pieces={} , max_pieces={}",
+                curr_portion, portion_num, start_portion_num, hints.is_contiguous, out_blkid.num_pieces(), max_pieces);
+        }
+    } while (nblks_remain && (portion_num != start_portion_num) && (out_blkid.num_pieces() < max_pieces));
```

Comment:
is there a reason why you replace `out_blkid.has_room()` ==> `out_blkid.num_pieces()`? 
My understanding is caller should always access has_room(), and not try to understand pieces inside a blk.

Reply:
it is for  `is_contiguous`, if `is_contiguous` is set we only want one pieces. `has_room()` doesnt serve this purpose 
```
auto const max_pieces = hints.is_contiguous ? 1u : MultiBlkId::max_pieces;
```

Reply:
no, num_pieces() means the existing allocated number of pieces on this blk id, technically if there is any blk already in this out_blkid (since it is a MultiBlkId), it will always return false (since max_pieces is 1 when is_contigous is true), right?

If is_contigous is true, we only need 1 pieces, I didn't get why has_room() doesn't serve the purpose, isn't it means out_blkid at least has 1 piece left?

Reply:
in general,  if `is_contigous == true` and we get a partial allocation , where the `num_pieces() == 1` and `has_room == true`,  we should NOT try other portion,  because even we can fulfill in other portion it is still not contiguous.  Through this will not happen as we allocate from bitmap we require all-or-none, but still , `num_pieces() < max_pieces` is more correct compared to `has_room  == true`

in contract, if ``is_contigous == false`, we should try other portion to fulfill the `nblks_remain`.

Reply:
Let's say in replication case, originator wants to send a MultiBlkId to follower, which contains a bunch of BlkIds. 
1st allocation is non-continous and saved to this mbid, and 2nd one is a contigous allocation. 

When originator allocates this alloc down to here, it already contains a few pieces, and this check will be false always (out_blkid.num_pieces() < max_pieces), we will not be able to allocate anything from anywhere, right? 

Have I missed anything?

Reply:
OK if we have a case that a non-empty `out_blkid` passed in,  you are right. Do we do it this way?  I was based on the assumption that `out_blkid` is an out opt.

Now this allocator is mostly on metasvc not in replication flow. But if you are talking nublox 2.0 or something that might be true.

Reply:
yes, I think we should still allow caller to such thing and in nublox 2.x replication, there should be such case.
If we don't allow caller to do such thing (out_blkid has pieces and is_contingous=true), my suggestion is to specifically assert that out_blkid.num_pieces() == 0 if is_contigous = true.

Reply:
I will add an assert and put the discussion context in the code.   So that later in Nublox 2.0 we can easily find out and based on the situation at that time we change the implementation.